runner/websocket.go

```diff
@@ -0,0 +1,132 @@
+package runner
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pganalyze/collector/output/pganalyze_collector"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+	"golang.org/x/net/http/httpproxy"
+	"google.golang.org/protobuf/proto"
+)
+
+func SetupWebsocketForAllServers(ctx context.Context, servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx := range servers {
+		go func(server *state.Server) {
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			for {
+				if server.WebSocket.Load() == nil {
+					connect(ctx, server, globalCollectionOpts, logger)
+				}
+				time.Sleep(3 * 60 * time.Second) // Delay between reconnect attempts
+			}
+		}(servers[idx])
+	}
+}
+
+func connect(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	connCtx, cancelConn := context.WithCancel(ctx)
+	proxyConfig := httpproxy.Config{
+		HTTPProxy:  server.Config.HTTPProxy,
+		HTTPSProxy: server.Config.HTTPSProxy,
+		NoProxy:    server.Config.NoProxy,
+	}
+	dialer := websocket.Dialer{
+		ReadBufferSize:  10240,
+		WriteBufferSize: 10240,
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(req.URL)
+		},
+	}
+	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/snapshots/websocket")
+	if url.Scheme == "http" {
+		url.Scheme = "ws"
+	} else {
+		url.Scheme = "wss"
+	}
+	headers := make(map[string][]string)
+	headers["Pganalyze-Api-Key"] = []string{server.Config.APIKey}
+	headers["Pganalyze-System-Id"] = []string{server.Config.SystemID}
+	headers["Pganalyze-System-Type"] = []string{server.Config.SystemType}
+	headers["Pganalyze-System-Scope"] = []string{server.Config.SystemScope}
+	headers["Pganalyze-System-Id-Fallback"] = []string{server.Config.SystemIDFallback}
+	headers["Pganalyze-System-Type-Fallback"] = []string{server.Config.SystemTypeFallback}
+	headers["Pganalyze-System-Scope-Fallback"] = []string{server.Config.SystemScopeFallback}
+	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
+	conn, response, err := dialer.DialContext(connCtx, url.String(), headers)
```

Comment:
I got most of the way through this, but ran into a circular dependency error:
```
package github.com/pganalyze/collector
    imports github.com/pganalyze/collector/runner
    imports github.com/pganalyze/collector/grant
    imports github.com/pganalyze/collector/runner: import cycle not allowed
```

In general I think this should be a separate PR, since it's solidly a refactor that's unrelated to the changes in this PR.

<details><summary>Full diff</summary>

```sql

diff --git a/config/read.go b/config/read.go
index 0a6b2d60..733521c5 100644
--- a/config/read.go
+++ b/config/read.go
@@ -2,10 +2,8 @@ package config
 
 import (
 	"context"
-	"crypto/tls"
 	"fmt"
 	"io/ioutil"
-	"net"
 	"net/http"
 	"net/url"
 	"os"
@@ -21,11 +19,8 @@ import (
 	sdkresource "go.opentelemetry.io/otel/sdk/resource"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
 	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
-	"golang.org/x/net/http/httpproxy"
 
 	"github.com/pganalyze/collector/util"
-
-	"github.com/hashicorp/go-retryablehttp"
 )
 
 const DefaultAPIBaseURL = "https://api.pganalyze.com"
@@ -372,70 +367,6 @@ func getDefaultConfig() *ServerConfig {
 	return config
 }
 
-func CreateHTTPClient(conf ServerConfig, logger *util.Logger, retry bool) *http.Client {
-	requireSSL := conf.APIBaseURL == DefaultAPIBaseURL
-	proxyConfig := httpproxy.Config{
-		HTTPProxy:  conf.HTTPProxy,
-		HTTPSProxy: conf.HTTPSProxy,
-		NoProxy:    conf.NoProxy,
-	}
-	transport := &http.Transport{
-		Proxy: func(req *http.Request) (*url.URL, error) {
-			return proxyConfig.ProxyFunc()(req.URL)
-		},
-		DialContext: (&net.Dialer{
-			Timeout:   30 * time.Second,
-			KeepAlive: 30 * time.Second,
-			DualStack: true,
-		}).DialContext,
-		MaxIdleConns:          100,
-		IdleConnTimeout:       90 * time.Second,
-		TLSHandshakeTimeout:   10 * time.Second,
-		ExpectContinueTimeout: 1 * time.Second,
-	}
-	if requireSSL {
-		transport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
-			matchesProxyURL := false
-			if proxyConfig.HTTPProxy != "" {
-				proxyURL, err := url.Parse(proxyConfig.HTTPProxy)
-				if err == nil && proxyURL.Host == addr {
-					matchesProxyURL = true
-				}
-			}
-			if proxyConfig.HTTPSProxy != "" {
-				proxyURL, err := url.Parse(proxyConfig.HTTPSProxy)
-				if err == nil && proxyURL.Host == addr {
-					matchesProxyURL = true
-				}
-			}
-			// Require secure conection for everything except proxies
-			if !matchesProxyURL && !strings.HasSuffix(addr, ":443") {
-				return nil, fmt.Errorf("Unencrypted connection is not permitted by pganalyze configuration")
-			}
-			return (&net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second, DualStack: true}).DialContext(ctx, network, addr)
-		}
-		transport.TLSClientConfig = &tls.Config{MinVersion: tls.VersionTLS12}
-	}
-
-	if retry {
-		client := retryablehttp.NewClient()
-		client.RetryWaitMin = 1 * time.Second
-		client.RetryWaitMax = 30 * time.Second
-		client.RetryMax = 4
-		client.Logger = nil
-		client.HTTPClient.Timeout = 120 * time.Second
-		client.HTTPClient.Transport = transport
-		return client.StandardClient()
-		// Note: StandardClient() only acts as a passthrough, handing the request to
-		// retryablehttp.Client whose nested HTTP client ends up using our custom transport.
-	} else {
-		return &http.Client{
-			Timeout:   120 * time.Second,
-			Transport: transport,
-		}
-	}
-}
-
 // CreateEC2IMDSHTTPClient - Create HTTP client for EC2 instance meta data service (IMDS)
 func CreateEC2IMDSHTTPClient(conf ServerConfig) *http.Client {
 	// Match https://github.com/aws/aws-sdk-go/pull/3066
diff --git a/grant/default.go b/grant/default.go
index 2d6eab66..534cfe76 100644
--- a/grant/default.go
+++ b/grant/default.go
@@ -10,6 +10,7 @@ import (
 	"github.com/pganalyze/collector/output/pganalyze_collector"
 	"github.com/pganalyze/collector/state"
 	"github.com/pganalyze/collector/util"
+	"github.com/pganalyze/collector/runner"
 )
 
 func GetDefaultGrant(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) (state.Grant, error) {
@@ -20,6 +21,8 @@ func GetDefaultGrant(ctx context.Context, server *state.Server, globalCollection
 		return grant, err
 	}
 
+	req.Header = runner.AuthHeaders(server.Config, globalCollectionOpts)
+
 	req.Header.Set("Pganalyze-Api-Key", server.Config.APIKey)
 	req.Header.Set("Pganalyze-System-Id", server.Config.SystemID)
 	req.Header.Set("Pganalyze-System-Type", server.Config.SystemType)
diff --git a/logs/parse_test.go b/logs/parse_test.go
index f9a68566..fb72a7c4 100644
--- a/logs/parse_test.go
+++ b/logs/parse_test.go
@@ -129,7 +129,7 @@ var parseTests = []parseTestpair{
 		"2018-09-27 06:57:01.030 EST [20194][] : [1-1] [app=pganalyze_collector] LOG:  connection received: host=[local]",
 		nil,
 		state.LogLine{
-			OccurredAt:    time.Date(2018, time.September, 27, 6, 57, 1, 30*1000*1000, time.FixedZone("EST", -5*3600)),
+			OccurredAt:    time.Date(2018, time.September, 27, 6, 57, 1, 30*1000*1000, time.FixedZone("CDT", -5*3600)),
 			LogLevel:      pganalyze_collector.LogLineInformation_LOG,
 			BackendPid:    20194,
 			LogLineNumber: 1,
diff --git a/main.go b/main.go
index d50c7135..7e6976d8 100644
--- a/main.go
+++ b/main.go
@@ -60,8 +60,8 @@ func run(ctx context.Context, wg *sync.WaitGroup, globalCollectionOpts state.Col
 		prefixedLogger := logger.WithPrefix(cfg.SectionName)
 		prefixedLogger.PrintVerbose("Identified as api_system_type: %s, api_system_scope: %s, api_system_id: %s", cfg.SystemType, cfg.SystemScope, cfg.SystemID)
 
-		conf.Servers[idx].HTTPClient = config.CreateHTTPClient(cfg, prefixedLogger, false)
-		conf.Servers[idx].HTTPClientWithRetry = config.CreateHTTPClient(cfg, prefixedLogger, true)
+		conf.Servers[idx].HTTPClient = runner.CreateHTTPClient(cfg, prefixedLogger, false)
+		conf.Servers[idx].HTTPClientWithRetry = runner.CreateHTTPClient(cfg, prefixedLogger, true)
 		if cfg.OtelExporterOtlpEndpoint != "" {
 			conf.Servers[idx].OTelTracingProvider, conf.Servers[idx].OTelTracingProviderShutdownFunc, err = config.CreateOTelTracingProvider(ctx, cfg)
 			logger.PrintVerbose("Initializing OpenTelemetry tracing provider with endpoint: %s", cfg.OtelExporterOtlpEndpoint)
diff --git a/runner/util.go b/runner/util.go
new file mode 100644
index 00000000..b0c589d9
--- /dev/null
+++ b/runner/util.go
@@ -0,0 +1,125 @@
+package runner
+
+import (
+	"context"
+	"net"
+	"time"
+	"net/http"
+	"net/url"
+	"crypto/tls"
+	"fmt"
+	"strings"
+
+	"github.com/gorilla/websocket"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+	"github.com/pganalyze/collector/config"
+	"golang.org/x/net/http/httpproxy"
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+func CreateHTTPClient(conf config.ServerConfig, logger *util.Logger, retry bool) *http.Client {
+	requireSSL := conf.APIBaseURL == config.DefaultAPIBaseURL
+	proxyConfig := httpproxy.Config{
+		HTTPProxy:  conf.HTTPProxy,
+		HTTPSProxy: conf.HTTPSProxy,
+		NoProxy:    conf.NoProxy,
+	}
+	transport := &http.Transport{
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(req.URL)
+		},
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+	if requireSSL {
+		transport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
+			matchesProxyURL := false
+			if proxyConfig.HTTPProxy != "" {
+				proxyURL, err := url.Parse(proxyConfig.HTTPProxy)
+				if err == nil && proxyURL.Host == addr {
+					matchesProxyURL = true
+				}
+			}
+			if proxyConfig.HTTPSProxy != "" {
+				proxyURL, err := url.Parse(proxyConfig.HTTPSProxy)
+				if err == nil && proxyURL.Host == addr {
+					matchesProxyURL = true
+				}
+			}
+			// Require secure conection for everything except proxies
+			if !matchesProxyURL && !strings.HasSuffix(addr, ":443") {
+				return nil, fmt.Errorf("Unencrypted connection is not permitted by pganalyze configuration")
+			}
+			return (&net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second, DualStack: true}).DialContext(ctx, network, addr)
+		}
+		transport.TLSClientConfig = &tls.Config{MinVersion: tls.VersionTLS12}
+	}
+
+	if retry {
+		client := retryablehttp.NewClient()
+		client.RetryWaitMin = 1 * time.Second
+		client.RetryWaitMax = 30 * time.Second
+		client.RetryMax = 4
+		client.Logger = nil
+		client.HTTPClient.Timeout = 120 * time.Second
+		client.HTTPClient.Transport = transport
+		return client.StandardClient()
+		// Note: StandardClient() only acts as a passthrough, handing the request to
+		// retryablehttp.Client whose nested HTTP client ends up using our custom transport.
+	} else {
+		return &http.Client{
+			Timeout:   120 * time.Second,
+			Transport: transport,
+		}
+	}
+}
+
+func CreateWebsocketConnection(ctx context.Context, conf config.ServerConfig, globalCollectionOpts state.CollectionOpts) (*websocket.Conn, *http.Response, error) {
+	proxyConfig := httpproxy.Config{
+		HTTPProxy:  conf.HTTPProxy,
+		HTTPSProxy: conf.HTTPSProxy,
+		NoProxy:    conf.NoProxy,
+	}
+	dialer := websocket.Dialer{
+		ReadBufferSize:  10240,
+		WriteBufferSize: 10240,
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(req.URL)
+		},
+	}
+	url, err := url.Parse(conf.APIBaseURL + "/v2/snapshots/websocket")
+	if err != nil {
+		return nil, nil, err
+	}
+	if url.Scheme == "http" {
+		url.Scheme = "ws"
+	} else {
+		url.Scheme = "wss"
+	}
+	headers := AuthHeaders(conf, globalCollectionOpts)
+	return dialer.DialContext(ctx, url.String(), headers)
+}
+
+func AuthHeaders(conf config.ServerConfig, globalCollectionOpts state.CollectionOpts) map[string][]string {
+	headers := make(map[string][]string)
+	headers["Pganalyze-Api-Key"] = []string{conf.APIKey}
+	headers["Pganalyze-System-Id"] = []string{conf.SystemID}
+	headers["Pganalyze-System-Type"] = []string{conf.SystemType}
+	headers["Pganalyze-System-Scope"] = []string{conf.SystemScope}
+	headers["Pganalyze-System-Id-Fallback"] = []string{conf.SystemIDFallback}
+	headers["Pganalyze-System-Type-Fallback"] = []string{conf.SystemTypeFallback}
+	headers["Pganalyze-System-Scope-Fallback"] = []string{conf.SystemScopeFallback}
+	if globalCollectionOpts.TestRun {
+		headers["Pganalyze-Test-Run"] = []string{"true"}
+	}
+	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
+	return headers
+}
diff --git a/runner/websocket.go b/runner/websocket.go
index 5febe66b..b7f1c0fd 100644
--- a/runner/websocket.go
+++ b/runner/websocket.go
@@ -5,15 +5,12 @@ import (
 	"compress/zlib"
 	"context"
 	"io"
-	"net/http"
-	"net/url"
 	"time"
 
 	"github.com/gorilla/websocket"
 	"github.com/pganalyze/collector/output/pganalyze_collector"
 	"github.com/pganalyze/collector/state"
 	"github.com/pganalyze/collector/util"
-	"golang.org/x/net/http/httpproxy"
 	"google.golang.org/protobuf/proto"
 )
 
@@ -38,34 +35,7 @@ func SetupWebsocketForAllServers(ctx context.Context, servers []*state.Server, g
 
 func connect(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
 	connCtx, cancelConn := context.WithCancel(ctx)
-	proxyConfig := httpproxy.Config{
-		HTTPProxy:  server.Config.HTTPProxy,
-		HTTPSProxy: server.Config.HTTPSProxy,
-		NoProxy:    server.Config.NoProxy,
-	}
-	dialer := websocket.Dialer{
-		ReadBufferSize:  10240,
-		WriteBufferSize: 10240,
-		Proxy: func(req *http.Request) (*url.URL, error) {
-			return proxyConfig.ProxyFunc()(req.URL)
-		},
-	}
-	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/snapshots/websocket")
-	if url.Scheme == "http" {
-		url.Scheme = "ws"
-	} else {
-		url.Scheme = "wss"
-	}
-	headers := make(map[string][]string)
-	headers["Pganalyze-Api-Key"] = []string{server.Config.APIKey}
-	headers["Pganalyze-System-Id"] = []string{server.Config.SystemID}
-	headers["Pganalyze-System-Type"] = []string{server.Config.SystemType}
-	headers["Pganalyze-System-Scope"] = []string{server.Config.SystemScope}
-	headers["Pganalyze-System-Id-Fallback"] = []string{server.Config.SystemIDFallback}
-	headers["Pganalyze-System-Type-Fallback"] = []string{server.Config.SystemTypeFallback}
-	headers["Pganalyze-System-Scope-Fallback"] = []string{server.Config.SystemScopeFallback}
-	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
-	conn, response, err := dialer.DialContext(connCtx, url.String(), headers)
+	conn, response, err := CreateWebsocketConnection(connCtx, server.Config, globalCollectionOpts)
 	if err != nil {
 		cancelConn()
 		logger.PrintWarning("Error starting websocket: %s %v", err, response)

Reply:
Meh, that's annoying re: circular dependency. Agreed we can do this in a separate PR.